src/tunnel.cc

```diff
@@ -168,6 +168,8 @@ class TunnelStateData: public PeerSelectionInitiator
     HappyConnOpenerPointer connOpener; ///< current connection opening job
     ResolvedPeersPointer destinations; ///< paths for forwarding the request
     bool destinationsFound; ///< At least one candidate path found
+    // TODO: remove after fixing deferred reads in TunnelStateData::copyRead()
```

Comment:
I don't think this comment is very clear what is needed to fix the TODO.
 What is eg "proper" ? just _doing_ deferred reads at all ?

Reply:
The point is that once we fix the tunnel delay code with DeferredRead, removing the misused event, there should be no need to keep the context TunnelStateData because there should be no context loss when delaying reads.

Reply:
> I don't think this comment is very clear what is needed to fix the TODO.

@yadij, please reread the TODO comment. It is very clear about what needs to be done here -- the data member should be removed. The comment also provides a precondition for that removal -- fixing a bug in TunnelStateData::copyRead(). The details of that bug are documented in that method. They should not be documented _here_, in this TODO.

>  What is eg "proper" ? just doing deferred reads at all ?

This TODO does not use the word "proper". You are probably referring to the PR description, which contained the following sentence:

> Future adjusting this scope by proper DeferredRead operations should eliminate events thus making TunnelStateData::codeContext unnecessary (TODO).

The wording can be polished and the word "proper" can be removed (now done), but this was a (minor) English problem, not a PR scope/implementation problem. Please see if the adjusted [wording](https://github.com/squid-cache/squid/pull/560#issue-380369669) addresses your concern.

Reply:
The new wording is clear. Thank you.